File: whisper_lyrics/exporters.py
"""Exporters for synchronized lyrics (LRC, Enhanced Karaoke LRC),
Subtitles (SRT, WebVTT), and structured JSON.
"""

from typing import Dict, Any, List


def format_lrc_timestamp(seconds: float) -> str:
    """Format seconds into LRC timestamp [mm:ss.xx] (hundredths of a second)."""
    if seconds < 0:
        seconds = 0.0
    minutes = int(seconds // 60)
    remaining_seconds = seconds % 60
    hundredths = int((remaining_seconds - int(remaining_seconds)) * 100)
    secs = int(remaining_seconds)
    return f"{minutes:02d}:{secs:02d}.{hundredths:02d}"


def format_srt_timestamp(seconds: float) -> str:
    """Format seconds into SRT timestamp HH:MM:SS,mmm."""
    if seconds < 0:
        seconds = 0.0
    hours = int(seconds // 3600)
    minutes = int((seconds % 3600) // 60)
    secs = int(seconds % 60)
    milliseconds = int((seconds - int(seconds)) * 1000)
    return f"{hours:02d}:{minutes:02d}:{secs:02d},{milliseconds:03d}"


def format_vtt_timestamp(seconds: float) -> str:
    """Format seconds into WebVTT timestamp HH:MM:SS.mmm."""
    if seconds < 0:
        seconds = 0.0
    hours = int(seconds // 3600)
    minutes = int((seconds % 3600) // 60)
    secs = int(seconds % 60)
    milliseconds = int((seconds - int(seconds)) * 1000)
    return f"{hours:02d}:{minutes:02d}:{secs:02d}.{milliseconds:03d}"


def to_lrc(result: Dict[str, Any], title: str = "", artist: str = "") -> str:
    """
    Generate standard line-by-line synchronized LRC lyrics.
    Format: [mm:ss.xx] Lyric line
    """
    lines: List[str] = []
    
    # Optional metadata headers
    if title:
        lines.append(f"[ti:{title}]")
    if artist:
        lines.append(f"[ar:{artist}]")
    lines.append("[by:Whisper-XPU-Vivestream]")
    
    segments = result.get("segments", [])
    for seg in segments:
        text = seg.get("text", "").strip()
        if not text:
            continue
        start_time = seg.get("start", 0.0)
        ts = format_lrc_timestamp(start_time)
        lines.append(f"[{ts}] {text}")

    return "\n".join(lines) + "\n"


def to_enhanced_lrc(result: Dict[str, Any], title: str = "", artist: str = "") -> str:
    """
    Generate enhanced word-by-word Karaoke LRC lyrics.
    Format: [mm:ss.xx] <mm:ss.xx> Word1 <mm:ss.xx> Word2 ...
    Falls back to line-by-line if word timestamps are not present.
    """
    lines: List[str] = []

    if title:
        lines.append(f"[ti:{title}]")
    if artist:
        lines.append(f"[ar:{artist}]")
    lines.append("[by:Whisper-XPU-Vivestream-Enhanced]")

    segments = result.get("segments", [])
    for seg in segments:
        words = seg.get("words", [])
        start_time = seg.get("start", 0.0)
        seg_ts = format_lrc_timestamp(start_time)

        if words:
            word_parts = []
            for w in words:
                w_text = w.get("word", "").strip()
                if not w_text:
                    continue
                w_start = format_lrc_timestamp(w.get("start", start_time))
                word_parts.append(f"<{w_start}> {w_text}")
            
            if word_parts:
                lines.append(f"[{seg_ts}] " + " ".join(word_parts))
            else:
                text = seg.get("text", "").strip()
                if text:
                    lines.append(f"[{seg_ts}] {text}")
        else:
            text = seg.get("text", "").strip()
            if text:
                lines.append(f"[{seg_ts}] {text}")

    return "\n".join(lines) + "\n"


def to_srt(result: Dict[str, Any]) -> str:
    """Generate standard SubRip (.srt) subtitles."""
    blocks: List[str] = []
    segments = result.get("segments", [])

    idx = 1
    for seg in segments:
        text = seg.get("text", "").strip()
        if not text:
            continue
        start_str = format_srt_timestamp(seg.get("start", 0.0))
        end_str = format_srt_timestamp(seg.get("end", 0.0))
        blocks.append(f"{idx}\n{start_str} --> {end_str}\n{text}\n")
        idx += 1

    return "\n".join(blocks)


def to_vtt(result: Dict[str, Any]) -> str:
    """Generate WebVTT (.vtt) subtitles."""
    lines: List[str] = ["WEBVTT", ""]
    segments = result.get("segments", [])

    for seg in segments:
        text = seg.get("text", "").strip()
        if not text:
            continue
        start_str = format_vtt_timestamp(seg.get("start", 0.0))
        end_str = format_vtt_timestamp(seg.get("end", 0.0))
        lines.append(f"{start_str} --> {end_str}")
        lines.append(text)
        lines.append("")

    return "\n".join(lines)


def to_json(result: Dict[str, Any]) -> Dict[str, Any]:
    """Return cleaned, structured dictionary suitable for JSON serialization in Tauri/Rust."""
    cleaned_segments = []
    for s in result.get("segments", []):
        words = []
        for w in s.get("words", []):
            words.append({
                "word": w.get("word", ""),
                "start": round(w.get("start", 0.0), 3),
                "end": round(w.get("end", 0.0), 3),
                "probability": round(w.get("probability", 1.0), 3),
            })
        cleaned_segments.append({
            "id": s.get("id", 0),
            "start": round(s.get("start", 0.0), 3),
            "end": round(s.get("end", 0.0), 3),
            "text": s.get("text", "").strip(),
            "words": words,
        })

    return {
        "text": result.get("text", "").strip(),
        "language": result.get("language", ""),
        "segments": cleaned_segments,
    }

File: vivestream-whisper/src/exporters.rs
//! Exporters for synchronized lyrics (LRC, Enhanced Karaoke LRC), subtitles (SRT, WebVTT), and JSON.

use serde::Serialize;
use std::fs::File;
use std::io::Write;
use std::path::Path;

#[derive(Debug, Serialize, Clone)]
pub struct WordTiming {
    pub word: String,
    pub start: f64,
    pub end: f64,
    pub probability: f64,
}

#[derive(Debug, Serialize, Clone)]
pub struct Segment {
    pub id: usize,
    pub start: f64,
    pub end: f64,
    pub text: String,
    pub words: Vec<WordTiming>,
}

#[derive(Debug, Serialize, Clone)]
pub struct TranscriptionResult {
    pub text: String,
    pub language: String,
    pub duration: f64,
    pub segments: Vec<Segment>,
}

pub fn format_lrc_timestamp(seconds: f64) -> String {
    let s = if seconds < 0.0 { 0.0 } else { seconds };
    let minutes = (s / 60.0) as u32;
    let rem = s % 60.0;
    let secs = rem as u32;
    let hundredths = ((rem - (secs as f64)) * 100.0) as u32;
    format!("{:02}:{:02}.{:02}", minutes, secs, hundredths)
}

pub fn format_srt_timestamp(seconds: f64) -> String {
    let s = if seconds < 0.0 { 0.0 } else { seconds };
    let hours = (s / 3600.0) as u32;
    let minutes = ((s % 3600.0) / 60.0) as u32;
    let secs = (s % 60.0) as u32;
    let millis = ((s - (s as u64 as f64)) * 1000.0) as u32;
    format!("{:02}:{:02}:{:02},{:03}", hours, minutes, secs, millis)
}

pub fn format_vtt_timestamp(seconds: f64) -> String {
    let s = if seconds < 0.0 { 0.0 } else { seconds };
    let hours = (s / 3600.0) as u32;
    let minutes = ((s % 3600.0) / 60.0) as u32;
    let secs = (s % 60.0) as u32;
    let millis = ((s - (s as u64 as f64)) * 1000.0) as u32;
    format!("{:02}:{:02}:{:02}.{:03}", hours, minutes, secs, millis)
}

pub fn to_lrc(result: &TranscriptionResult, title: &str, artist: &str) -> String {
    let mut out = String::new();
    if !title.is_empty() {
        out.push_str(&format!("[ti:{}]\n", title));
    }
    if !artist.is_empty() {
        out.push_str(&format!("[ar:{}]\n", artist));
    }
    out.push_str("[by:Vivestream-Revived-Native]\n");

    for seg in &result.segments {
        let text = seg.text.trim();
        if text.is_empty() {
            continue;
        }
        let ts = format_lrc_timestamp(seg.start);
        out.push_str(&format!("[{}] {}\n", ts, text));
    }
    out
}

pub fn to_enhanced_lrc(result: &TranscriptionResult, title: &str, artist: &str) -> String {
    let mut out = String::new();
    if !title.is_empty() {
        out.push_str(&format!("[ti:{}]\n", title));
    }
    if !artist.is_empty() {
        out.push_str(&format!("[ar:{}]\n", artist));
    }
    out.push_str("[by:Vivestream-Revived-Karaoke]\n");

    for seg in &result.segments {
        let seg_ts = format_lrc_timestamp(seg.start);
        if !seg.words.is_empty() {
            let mut word_parts = Vec::new();
            for w in &seg.words {
                let w_text = w.word.trim();
                if w_text.is_empty() {
                    continue;
                }
                let w_ts = format_lrc_timestamp(w.start);
                word_parts.push(format!("<{}> {}", w_ts, w_text));
            }
            if !word_parts.is_empty() {
                out.push_str(&format!("[{}] {}\n", seg_ts, word_parts.join(" ")));
            } else if !seg.text.trim().is_empty() {
                out.push_str(&format!("[{}] {}\n", seg_ts, seg.text.trim()));
            }
        } else if !seg.text.trim().is_empty() {
            out.push_str(&format!("[{}] {}\n", seg_ts, seg.text.trim()));
        }
    }
    out
}

pub fn to_srt(result: &TranscriptionResult) -> String {
    let mut out = String::new();
    let mut idx = 1;
    for seg in &result.segments {
        let text = seg.text.trim();
        if text.is_empty() {
            continue;
        }
        let start_str = format_srt_timestamp(seg.start);
        let end_str = format_srt_timestamp(seg.end);
        out.push_str(&format!("{}\n{} --> {}\n{}\n\n", idx, start_str, end_str, text));
        idx += 1;
    }
    out
}

pub fn to_vtt(result: &TranscriptionResult) -> String {
    let mut out = String::from("WEBVTT\n\n");
    for seg in &result.segments {
        let text = seg.text.trim();
        if text.is_empty() {
            continue;
        }
        let start_str = format_vtt_timestamp(seg.start);
        let end_str = format_vtt_timestamp(seg.end);
        out.push_str(&format!("{} --> {}\n{}\n\n", start_str, end_str, text));
    }
    out
}

pub fn save_files(
    result: &TranscriptionResult,
    prefix_path: &Path,
    formats: &[String],
    title: &str,
    artist: &str,
) -> std::io::Result<Vec<(String, String)>> {
    let mut written = Vec::new();

    if let Some(parent) = prefix_path.parent() {
        std::fs::create_dir_all(parent)?;
    }

    for fmt in formats {
        let f_lower = fmt.to_lowercase();
        match f_lower.as_str() {
            "lrc" => {
                let path = format!("{}.lrc", prefix_path.display());
                let content = to_lrc(result, title, artist);
                File::create(&path)?.write_all(content.as_bytes())?;
                written.push(("lrc".to_string(), path));
            }
            "elrc" => {
                let path = format!("{}.enhanced.lrc", prefix_path.display());
                let content = to_enhanced_lrc(result, title, artist);
                File::create(&path)?.write_all(content.as_bytes())?;
                written.push(("elrc".to_string(), path));
            }
            "srt" => {
                let path = format!("{}.srt", prefix_path.display());
                let content = to_srt(result);
                File::create(&path)?.write_all(content.as_bytes())?;
                written.push(("srt".to_string(), path));
            }
            "vtt" => {
                let path = format!("{}.vtt", prefix_path.display());
                let content = to_vtt(result);
                File::create(&path)?.write_all(content.as_bytes())?;
                written.push(("vtt".to_string(), path));
            }
            "json" => {
                let path = format!("{}.json", prefix_path.display());
                let content = serde_json::to_string_pretty(result)?;
                File::create(&path)?.write_all(content.as_bytes())?;
                written.push(("json".to_string(), path));
            }
            _ => {}
        }
    }
    Ok(written)
}
